kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/credentials/Ec2CredentialsProvider.java

```diff
@@ -0,0 +1,285 @@
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Provider that obtains {@link Credentials} from the metadata server of the EC2 instance.
+ * <p>
+ * The provider will keep returning the same credential until the credential reaches
+ * a configured factor of the credential's lifespan.  At which point, a preemptive
+ * background refresh of the credential will be performed.  Until the refresh is complete
+ * the caller will continue to receive the existing credential.  Once the refresh is complete
+ * subsequent calls will see the updated credential.
+ * </p>
+ * <p>
+ * If an error occurs whilst retrieving the credential, the next call will cause the
+ * provider to try again.  A progress backoff is applied to retry attempts.
+ * </p>
+ */
+public class Ec2CredentialsProvider implements CredentialsProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(Ec2CredentialsProvider.class);
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().registerModule(new JavaTimeModule());
+
+    private static final TypeReference<SecurityCredentials> SECURITY_CREDENTIALS_RESPONSE_TYPE_REF = new TypeReference<>() {
+    };
+    @SuppressWarnings("java:S1313")
+    private static final URI DEFAULT_IP4_METADATA_ENDPOINT = URI.create("http://169.254.169.254/");
+    private static final Duration HTTP_REQUEST_TIMEOUT = Duration.ofSeconds(10);
+    private static final Duration HTTP_CONNECT_TIMEOUT = Duration.ofSeconds(10);
+    private static final String AWS_METADATA_TOKEN_TTL_SECONDS_HEADER = "X-aws-ec2-metadata-token-ttl-seconds";
+    private static final String AWS_METADATA_TOKEN_HEADER = "X-aws-ec2-metadata-token";
+    private static final String AWS_TOKEN_EXPIRATION_SECONDS = "60";
+    private static final double DEFAULT_CREDENTIALS_LIFETIME_FACTOR = 0.80;
+
+    /**
+     * EC2 Token Retrieval Endpoint.
+     * Note that  use of "latest" is
+     * <a href="https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#imds-considerations">required</a>.
+     */
+    static final String TOKEN_RETRIEVAL_ENDPOINT = "/latest/api/token";
+
+    /**
+     * EC2 Meta-data security credentials endpoint.
+     */
+    static final String META_DATA_IAM_SECURITY_CREDENTIALS_ENDPOINT = "/2024-04-11/meta-data/iam/security-credentials/";
+
+    private final Clock systemClock;
+    private final AtomicReference<CompletableFuture<SecurityCredentials>> current = new AtomicReference<>();
+
+    private final AtomicLong tokenRefreshErrorCount = new AtomicLong();
+    private final Ec2CredentialsProviderConfig config;
+    private final HttpClient client;
+
+    private final ScheduledExecutorService executorService;
+    @SuppressWarnings({ "java:S2245", "java:S2119" }) // Random used for backoff jitter, it does not need to be securely random.
+    private final ExponentialBackoff backoff = new ExponentialBackoff(500, 2, 60000, new Random().nextDouble());
```

Comment:
The motivation for jitter was to prevent a scaled out replicas of Kroxylicious all retrying in lockstep.   It is probably unlikely they would actually fall into localstep, but this still feels like the right approach.  I'd prefer to keep the `Random()`.